src/shims/sync.rs

```diff
@@ -0,0 +1,519 @@
+use rustc_middle::ty::{TyKind, TypeAndMut};
+use rustc_target::abi::{LayoutOf, Size};
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+fn assert_ptr_target_min_size<'mir, 'tcx: 'mir>(
+    ecx: &MiriEvalContext<'mir, 'tcx>,
+    operand: OpTy<'tcx, Tag>,
+    min_size: u64,
+) -> InterpResult<'tcx, ()> {
+    let target_ty = match operand.layout.ty.kind {
+        TyKind::RawPtr(TypeAndMut { ty, mutbl: _ }) => ty,
+        _ => panic!("Argument to pthread function was not a raw pointer"),
+    };
+    let target_layout = ecx.layout_of(target_ty)?;
+    assert!(target_layout.size.bytes() >= min_size);
+    Ok(())
+}
+
+// pthread_mutexattr_t is either 4 or 8 bytes, depending on the platform.
+
+// Our chosen memory layout: store an i32 in the first four bytes equal to the
+// corresponding libc mutex kind constant (i.e. PTHREAD_MUTEX_NORMAL)
+
+fn mutexattr_get_kind<'mir, 'tcx: 'mir>(
+    ecx: &MiriEvalContext<'mir, 'tcx>,
+    attr_op: OpTy<'tcx, Tag>,
+) -> InterpResult<'tcx, ScalarMaybeUndef<Tag>> {
+    // Ensure that the following read at an offset to the attr pointer is within bounds
+    assert_ptr_target_min_size(ecx, attr_op, 4)?;
+    let attr_place = ecx.deref_operand(attr_op)?;
+    let kind_place = attr_place.offset(Size::ZERO, MemPlaceMeta::None, ecx.i32_layout()?, ecx)?;
```

Comment:
Hm okay, looking at this, `ecx.machine.layouts.i32` would make things long enough to warrant wrapping the call over multiple lines. But it also would not be terrible.

I am undecided about which way is better.

Reply:
Part of my rationale for having the wrapper was that I otherwise had to pass the method ecx as an argument as well, i.e. `ecx.machine.layouts.i32(ecx)?`, which was a minor annoyance.

Reply:
I agree that with the lazy approach, the wrapper makes sense.
I am not convinced the lazy approach is necessary, but I am willing to go with it for now.